### PR TITLE
WIP code cleanup + event-handling opinions

### PR DESCRIPTION
**Pull request by kjvic**
## Major changes
Note: this PR is very much a WIP, but there are some decent bits that are worth checking out. Those bits are listed below (everything else you should pretty much ignore).

### **ShootBallMechanic**
 combined into BallCarrier

### **BallCarrier**
* Note: the function `StopCarrying()` is responsible for doing cleanup that is *internal to BallCarrier*. It should be called whenever you stop carrying -- stolen from, drop ball (i.e. shot timer times out), shoot (i.e. charging + release button), etc.
* Likewise, `StartCarryingBall` handles all internal logic (that's common to any kind of carry -- `steal`, `carry`, etc.).
* Note: events like `onStoppedCarryingBall` are ONLY called *after* the BallCarrier has actually finished doing all of its internal stop-carrying-ball cleanup. This guarantees that observer components (e.g. slowmo on possession) will never be notified if the BallCarrier is in a half-baked state.
* Note: I added a function `IsInNullZone` to BallCarrier.
    * It could probably be in its own component, but that seemed like overkill for now.
    * Right now, the only place where it's used is in `Ball` (the `Ball` only updates the `BallActivation` if `BallCarrier.IsInNullZone() == false` )

### **Ball**
Ball is now very simple. It keeps track of its owner and has a reference to BallActivation, and basically nothing else.
* The ball tells the BallActivation to update whenever the ball is carried, but the real color-change/passing/goal-switching/can-score logic is either handled in BallActivation, or in components that subscribe to related events (see next bullet)

### **BallActivation**
encapsulates the goal-switching logic (i.e. need to pass in order to fully "activate" the ball).
* When the `Ball` is carried, it tells the `BallActivation` to update itself based on who touched it. `BallActivation` may then fire an event like `OnBallDominated` (e.g. if the ball is pink & filled/can score a goal).
* Rather than *querying the ball* to check if a collision should count as a goal, the `Goal` will listen for the events `OnBallDominated` and `OnBallNeutralized` and keep track of whether the ball is eligible to score.

### **Goal**
refactored to work with `BallActivation`

### **PlayerDashMechanic**
* Only partially refactored
* See function StunAndSteal
(note: the syntax `if (ballCarrier.CanSteal(ball)) { ballCarrier.Steal(ball); }` -- much more maintainable to let BallCarrier determine if ball can be stolen, rather than computing `shouldSteal` as needed)

### **BlowbackOnPossession**
WIP, but the intention is to factor the "blowback on possession" behavior into its own component to demonstrate how events promote decoupling. 
* Note: `PlayerStun` is ultimately responsible for the actual stun logic
* => the `BlowbackOnPossession` component is *really* only there to encapsulate the fact that possession triggers 1. blowback/stun for enemy players 2. a blowback particle effect.
* In other words, `BlowbackOnPossession` basically says "Hey `PlayerStun`! Do your thing!". You *could* just add that logic to `PlayerStun`, but then you can't remove "BlowbackOnPossession" without removing "all of PlayerStun/Blowback" :(

### CODE REVIEW

File: Assets/Scripts/Effects/StunEffect.cs [hunk 1]
```diff
@@ -2,27 +2,40 @@
...
 {
     public float flashInterval = 0.1f;
     private bool stopEffect = false;
 
     private void Start()
     {
-        PlayerStateManager stateManager = this.EnsureComponent<PlayerStateManager>();
+        // PlayerStateManager stateManager = this.EnsureComponent<PlayerStateManager>();
+        EM.onPlayerStunned += (EM.onPlayerStunnedArgs args) => {
+            Player player = args.player;
```
Comment (kjvic) [inline Assets/Scripts/Effects/StunEffect.cs:15]: Note: ignore this usage of events. This is not actually how I intend to use them -- I was just porting existing semantics to use EventManager syntax.

Specifically, I would _**not**_ want the PlayerStun component to take action based on events like `onPlayerStunned`. Instead, I would want any `PlayerStun` to be the only component who's supposed to fire `onPlayerStunned`, and `PlayerStun` would do any internal stun-related logic in a private member function.

**Comment by kjvic** [comment]
:( I'm just gonna try to split this up and merge elsewhere.
